lib/uktest/include/uk/test.h

```diff
@@ -0,0 +1,595 @@
+	static int UK_TESTSUITE_FN(suite)(void)				\
+	{								\
+		uk_testsuite_add(&suite);				\
+		uk_testsuite_run(&suite);				\
+		return 0;						\
+	}								\
+	uk_initcall_class_prio(UK_TESTSUITE_FN(suite), class, prio)
+
+/**
+ * Add a test suite to be run in the "early" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_early_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_EARLY,   prio)
+
+/**
+ * Add a test suite to be run in the "plat" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_plat_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_PLAT,   prio)
+
+/**
+ * Add a test suite to be run in the "lib" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_lib_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_LIB,    prio)
+
+/**
+ * Add a test suite to be run in the "rootfs" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_rootfs_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_ROOTFS, prio)
+
+/**
+ * Add a test suite to be run in the "sys" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_sys_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_SYS,    prio)
+
+/**
+ * Add a test suite to be run in the "late" stage of the inittab.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_late_testsuite_prio(suite, prio) \
+	uk_test_at_initcall_prio(suite, UK_INIT_CLASS_LATE,   prio)
+
+/**
+ * The default registration for a test suite with a desird priority level.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ * @param prio
+ *   The priority of this suite at this stage.
+ */
+#define uk_testsuite_prio(suite, prio) \
+	uk_late_testsuite_prio(suite, prio)
+
+/**
+ * The default registration for a test suite.
+ *
+ * @param suite
+ *   The pointer to the suite to add this stage.
+ */
+#define uk_testsuite_register(suite) \
+	uk_late_testsuite_prio(suite, UK_PRIO_LATEST)
+
+static inline void
+_uk_test_do_assert(struct uk_testcase *esac, bool cond, const char *fmt, ...)
+{
+#ifdef CONFIG_LIBUKTEST_LOG_TESTS
+	va_list ap;
+	int pad_len;
+	char msg[UKT_COLWIDTH];
+
+	/* Save formatted message to buffer. */
+	va_start(ap, fmt);
+	vsprintf(msg, fmt, ap);
```

Comment:
In the unlikely case that the message (i.e., cond) is longer than UK_COLWIDTH this corrupts the stack. The comparison in line 332 is thus not sufficient. Suggest to use vsnprintf instead. Comparison can be removed.

```suggestion
	vsnprintf(msg, sizeof(msg), fmt, ap);
```